File: JsonAgent/VisualBasic/BitSefClient.vb
Imports System
Imports System.Net.Http
Imports System.Security.Cryptography
Imports System.Text
Imports System.Threading.Tasks
Public Class BitSefClient
 Implements IDisposable
 Private ReadOnly _baseUrl As String, _secret As String
 Private ReadOnly _http As New HttpClient()
 Public Sub New(baseUrl As String, sharedSecret As String) : _baseUrl=baseUrl.TrimEnd("/"c) : _secret=sharedSecret : End Sub
 Public Function FiscalJsonAsync(json As String) As Task(Of String) : Return PostAsync("/api/bitsef/fiscal",json,"application/json") : End Function
 Public Function CommandJsonAsync(json As String) As Task(Of String) : Return PostAsync("/api/bitsef/command",json,"application/json") : End Function
 Public Function CopyAsync(no As String) As Task(Of String) : Return PostAsync("/api/bitsef/copy/" & no,"",Nothing) : End Function
 Public Function PdfAsync(no As String) As Task(Of String) : Return PostAsync("/api/bitsef/pdf/" & no,"",Nothing) : End Function
 Public Function DirectCsvAsync(type As String,no As String,csv As String) As Task(Of String) : Return PostAsync("/api/bitsef/csv/" & type & "/" & no,csv,"text/plain") : End Function
 Public Async Function StatusAsync(id As String) As Task(Of String) : Dim r=Await _http.GetAsync(_baseUrl & "/api/bitsef/status/" & id) : Dim t=Await r.Content.ReadAsStringAsync() : r.EnsureSuccessStatusCode() : Return t : End Function
 Private Async Function PostAsync(path As String,body As String,mediaType As String) As Task(Of String)
  Dim ts=DateTimeOffset.UtcNow.ToUnixTimeSeconds().ToString(), nonce=Guid.NewGuid().ToString("N")
  Dim canonical="POST" & vbLf & path & vbLf & ts & vbLf & nonce & vbLf & Sha256Hex(If(body,""))
  Using req As New HttpRequestMessage(HttpMethod.Post,_baseUrl & path)
   req.Headers.Add("X-BIT-Timestamp",ts) : req.Headers.Add("X-BIT-Nonce",nonce) : req.Headers.Add("X-BIT-Signature",HmacHex(canonical,_secret))
   If mediaType IsNot Nothing Then req.Content=New StringContent(If(body,""),Encoding.UTF8,mediaType)
   Using res=Await _http.SendAsync(req) : Dim text=Await res.Content.ReadAsStringAsync() : If Not res.IsSuccessStatusCode Then Throw New Exception("BIT-SEF HTTP " & CInt(res.StatusCode) & ": " & text) : Return text : End Using
  End Using
 End Function
 Private Shared Function Sha256Hex(s As String) As String : Using h=SHA256.Create() : Return Hex(h.ComputeHash(Encoding.UTF8.GetBytes(s))) : End Using : End Function
 Private Shared Function HmacHex(s As String,k As String) As String : Using h=New HMACSHA256(Encoding.UTF8.GetBytes(k)) : Return Hex(h.ComputeHash(Encoding.UTF8.GetBytes(s))) : End Using : End Function
 Private Shared Function Hex(b As Byte()) As String : Dim sb As New StringBuilder() : For Each x In b : sb.Append(x.ToString("x2")) : Next : Return sb.ToString() : End Function
 Public Sub Dispose() Implements IDisposable.Dispose : _http.Dispose() : End Sub
End Class

File: JsonAgent/php/BitSefClient.php
<?php
class BitSefClient {
 private string $baseUrl; private string $secret;
 function __construct(string $u,string $s){$this->baseUrl=rtrim($u,'/');$this->secret=$s;}
 function fiscalJson(string $j):string{return $this->post('/api/bitsef/fiscal',$j,'application/json');}
 function commandJson(string $j):string{return $this->post('/api/bitsef/command',$j,'application/json');}
 function copy(string $n):string{return $this->post('/api/bitsef/copy/'.$n,'',null);}
 function pdf(string $n):string{return $this->post('/api/bitsef/pdf/'.$n,'',null);}
 function directCsv(string $t,string $n,string $c):string{return $this->post('/api/bitsef/csv/'.$t.'/'.$n,$c,'text/plain; charset=utf-8');}
 function status(string $id):string{return file_get_contents($this->baseUrl.'/api/bitsef/status/'.$id);}
 private function post(string $p,string $b,?string $ct):string{$ts=(string)time();$nonce=bin2hex(random_bytes(16));$canonical="POST\n$p\n$ts\n$nonce\n".hash('sha256',$b);$sig=hash_hmac('sha256',$canonical,$this->secret);$h=["X-BIT-Timestamp: $ts","X-BIT-Nonce: $nonce","X-BIT-Signature: $sig"];if($ct)$h[]="Content-Type: $ct";$ch=curl_init($this->baseUrl.$p);curl_setopt_array($ch,[CURLOPT_POST=>true,CURLOPT_RETURNTRANSFER=>true,CURLOPT_HTTPHEADER=>$h]);if($ct)curl_setopt($ch,CURLOPT_POSTFIELDS,$b);$r=curl_exec($ch);$code=curl_getinfo($ch,CURLINFO_HTTP_CODE);if($r===false)throw new Exception(curl_error($ch));curl_close($ch);if($code<200||$code>=300)throw new Exception("BIT-SEF HTTP $code: $r");return $r;}
}
